### PR TITLE
Normalize internationalized domain names when adding to CORS

### PR DESCRIPTION
**Pull request by pranavkm**
Fixes https://github.com/aspnet/Home/issues/3353

### CODE REVIEW

File: test/Microsoft.AspNetCore.Cors.Test/CorsPolicyBuilderTests.cs [hunk 1]
```diff
@@ -299,5 +299,98 @@ public void DisallowCredential_SetsSupportsCredentials_ToFalse()
             var corsPolicy = builder.Build();
             Assert.False(corsPolicy.SupportsCredentials);
         }
+
+        [Fact]
+        public void GetNormalizedOrigin_ReturnsLowerCasedValue_IfStringIsNotParseableAsUrl()
```
Comment (Tratcher) [inline test/Microsoft.AspNetCore.Cors.Test/CorsPolicyBuilderTests.cs]: If the string is not parsable as a URI then the casing is the least of their worries.

File: test/Microsoft.AspNetCore.Cors.Test/CorsPolicyBuilderTests.cs [hunk 2]
```diff
@@ -299,5 +299,98 @@ public void DisallowCredential_SetsSupportsCredentials_ToFalse()
...
+
+            // Assert
+            Assert.Equal("x:\\test", normalizedOrigin);
+        }
+
+        [Fact]
+        public void GetNormalizedOrigin_ReturnsLowerCasedValue_IsStringIsNetworkPath()
+        {
+            // Arrange
+            var origin = "File://MNT/Share/";
```
Comment (Tratcher) [inline test/Microsoft.AspNetCore.Cors.Test/CorsPolicyBuilderTests.cs]: How is this a relevant test case?

File: src/Microsoft.AspNetCore.Cors/Infrastructure/CorsPolicyBuilder.cs [hunk 3]
```diff
@@ -40,12 +40,36 @@ public CorsPolicyBuilder WithOrigins(params string[] origins)
...
+                !string.Equals(uri.IdnHost, uri.Host, StringComparison.Ordinal))
+            {
+                var builder = new UriBuilder(uri.Scheme.ToLowerInvariant(), uri.IdnHost.ToLowerInvariant());
+                if (!uri.IsDefaultPort)
+                {
+                    // Uri does not have a way to differentiate between a port value inferred by default (e.g. Port = 80 for http://www.example.com) and
+                    // a default port value that is specified (e.g. Port = 80 for http://www.example.com:80). Although the HTTP or FETCH spec does not say 
+                    // anything about including the default port as part of the Origin header, at the time of writing, browsers drop "default" port when navigating
+                    // and when sending the Origin header. All this goes to say, it appears OK to drop an explicitly specified port, 
+                    // if it is the default port when working with an IDN host.
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Cors/Infrastructure/CorsPolicyBuilder.cs]: This comment doesn't match the GetNormalizedOrigin_PreservesPort test results.

File: test/Microsoft.AspNetCore.Cors.Test/CorsPolicyBuilderTests.cs (hunk 1 above)
Comment (pranavkm) [reply test/Microsoft.AspNetCore.Cors.Test/CorsPolicyBuilderTests.cs]: I sort of wanted to make sure the behavior is no different (we don't throw for instance) from what it looks like today.

File: test/Microsoft.AspNetCore.Cors.Test/CorsPolicyBuilderTests.cs [hunk 4]
```diff
@@ -299,5 +299,85 @@ public void DisallowCredential_SetsSupportsCredentials_ToFalse()
             var corsPolicy = builder.Build();
             Assert.False(corsPolicy.SupportsCredentials);
         }
+
+        [Theory]
+        [InlineData(@"Some-String", "some-string")]
+        [InlineData(@"x:\\Test", "x:\\test")]
+        [InlineData("FTP://Some-url", "ftp://some-url")]
+        public void GetNormalizedOrigin_ReturnsLowerCasedValue_IfStringIsNotHttpOrHttpsUrl(string origin, string expected)
```
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Cors.Test/CorsPolicyBuilderTests.cs]: Better?

File: test/Microsoft.AspNetCore.Cors.Test/CorsPolicyBuilderTests.cs [hunk 5]
```diff
@@ -299,5 +299,85 @@ public void DisallowCredential_SetsSupportsCredentials_ToFalse()
...
+            var normalizedOrigin = CorsPolicyBuilder.GetNormalizedOrigin(origin);
+
+            // Assert
+            Assert.Equal("http://www.example.com", normalizedOrigin);
+        }
+
+        [Theory]
+        [InlineData("http://www.Example.com:80", "http://www.example.com:80")]
+        [InlineData("https://www.Example.com:8080", "https://www.example.com:8080")]
+        public void GetNormalizedOrigin_PreservesPort_ForNonIdnHosts(string origin, string expected)
```
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Cors.Test/CorsPolicyBuilderTests.cs]: Updated test name

Comment (Tratcher) [reply test/Microsoft.AspNetCore.Cors.Test/CorsPolicyBuilderTests.cs]: Having different default port behavior for IDNs and non-IDNS doesn't sound like it's actually going to work. From your description above it sounds like dropping the default port is what browsers expect.

Comment (pranavkm) [reply test/Microsoft.AspNetCore.Cors.Test/CorsPolicyBuilderTests.cs]: My concern was that it's a really large browser matrix to test to be 100% certain. It seemed less problematic with Ids since they weren't working to start with. But it'd be more frustrating if we broke it for apps \ browsers that worked correctly.

Comment (Tratcher) [reply test/Microsoft.AspNetCore.Cors.Test/CorsPolicyBuilderTests.cs]: You need to explain all of this normalization in the doc comments on WithOrigins.

Comment (pranavkm) [reply test/Microsoft.AspNetCore.Cors.Test/CorsPolicyBuilderTests.cs]: Added docs

File: src/Microsoft.AspNetCore.Cors/Infrastructure/CorsPolicyBuilder.cs [hunk 6]
```diff
@@ -36,6 +37,21 @@ public CorsPolicyBuilder(CorsPolicy policy)
         /// </summary>
         /// <param name="origins">The origins that are allowed.</param>
         /// <returns>The current policy builder.</returns>
+        /// <remarks>
+        /// This method normalizes the origin headers prior to adding it to <see cref="CorsPolicy.Origins"/> to match
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Cors/Infrastructure/CorsPolicyBuilder.cs]: origin headers -> origin values

File: src/Microsoft.AspNetCore.Cors/Infrastructure/CorsPolicyBuilder.cs [hunk 7]
```diff
@@ -36,6 +37,21 @@ public CorsPolicyBuilder(CorsPolicy policy)
...
+        /// the normalization performed by the browser on the value sent in the <c>ORIGIN</c> header.
+        /// <list type="bullet">
+        /// <item>
+        /// If the specified origin has an internationalized domain name (IDN), the punycoded value is used. If the origin
+        /// specifies a default port (e.g. 443 for HTTPS or 80 for HTTP), this will be dropped as part of normalization.
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Cors/Infrastructure/CorsPolicyBuilder.cs]: Except the default port is only dropped for IDNs

Comment (pranavkm) [reply src/Microsoft.AspNetCore.Cors/Infrastructure/CorsPolicyBuilder.cs]: It says so, doesn't it: 

> If the specified origin has an internationalized domain name (IDN)
